### PR TITLE
[ipgen, rv_plic]  Enable specific module_instance_name for multiple plic groups

### PR DESCRIPTION
**Pull request by hcindyl**
Current ipgen for rv_plic only allow a single rv_plic group, and
generate the `ip_autogen` based on the template name. Adding a field in
the parameter `module_instance_name` to allow more plic groups to be generated.

To support this, all the non-template files in `hw/ip_templates/rv_plic`
(except docs) are converted to mako template files.

It is a no-op for top_earlgrey, with the slight EOF newline change from
the generated files.

Signed-off-by: Cindy Liu <hcindyl@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
is the idea here you are trying to generate multiple rv_plic modules in your design and they each have slightly different hjson settings?

**Comment by tjaychen** [comment]
@rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.

**Comment by hcindyl** [comment]
> @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.

I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.

**Comment by tjaychen** [comment]
> > @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.
> 
> I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.

ah yes, let me clarify. I'm just wondering if we should allow ipgen to be passed an "instance name" directly instead of just suffix.

**Comment by hcindyl** [comment]
> > > @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.
> > 
> > 
> > I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.
> 
> ah yes, let me clarify. I'm just wondering if we should allow ipgen to be passed an "instance name" directly instead of just suffix.

Got it. Please let me know if you prefer to pass `instance_name` instead, and I can adjust the PR accordingly.

**Comment by tjaychen** [comment]
> > > > @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.
> > > 
> > > 
> > > I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.
> > 
> > 
> > ah yes, let me clarify. I'm just wondering if we should allow ipgen to be passed an "instance name" directly instead of just suffix.
> 
> Got it. Please let me know if you prefer to pass `instance_name` instead, and I can adjust the PR accordingly.

i feel like that would make more sense right? @rswarbrick what do you think?
@hcindyl thank you for shepherding this change.

**Comment by hcindyl** [comment]
> > > > > @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.
> > > > 
> > > > 
> > > > I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.
> > > 
> > > 
> > > ah yes, let me clarify. I'm just wondering if we should allow ipgen to be passed an "instance name" directly instead of just suffix.
> > 
> > 
> > Got it. Please let me know if you prefer to pass `instance_name` instead, and I can adjust the PR accordingly.
> 
> i feel like that would make more sense right? @rswarbrick what do you think? @hcindyl thank you for shepherding this change.

Glad to help, it will also allow us to create multiple module instances without forking out the templates.

However, it seems we need to have a different `instance_name` definition
https://github.com/lowRISC/opentitan/tree/2b8c7f5/util/topgen.py#L70

I would still prefer to add the new `instance_name` field under `params`, and start making the change with `rv_plic` with this PR. What do you think?

**Comment by tjaychen** [comment]
> > > > > > @rswarbrick does it make sense to enhance ipgen in general to generate multiple modules based on a template? we could just supply unique names for each module.
> > > > > 
> > > > > 
> > > > > I think if we extend the `module_suffix` support in this PR to `alert_handler` it will be sufficient to support such a feature.
> > > > 
> > > > 
> > > > ah yes, let me clarify. I'm just wondering if we should allow ipgen to be passed an "instance name" directly instead of just suffix.
> > > 
> > > 
> > > Got it. Please let me know if you prefer to pass `instance_name` instead, and I can adjust the PR accordingly.
> > 
> > 
> > i feel like that would make more sense right? @rswarbrick what do you think? @hcindyl thank you for shepherding this change.
> 
> Glad to help, it will also allow us to create multiple module instances without forking out the templates.
> 
> However, it seems we need to have a different `instance_name` definition https://github.com/lowRISC/opentitan/tree/2b8c7f5/util/topgen.py#L70
> 
> I would still prefer to add the new `instance_name` field under `params`, and start making the change with `rv_plic` with this PR. What do you think?

yeah that sounds good.
Are you guys using our `topgen.py` directly? Because try as we did..i think `topgen` is actually going to end up being very specific to each design.  And there are also some issues with `topgen` today.

**Comment by hcindyl** [comment]
> > Glad to help, it will also allow us to create multiple module instances without forking out the templates.
> > However, it seems we need to have a different `instance_name` definition https://github.com/lowRISC/opentitan/tree/2b8c7f5/util/topgen.py#L70
> > I would still prefer to add the new `instance_name` field under `params`, and start making the change with `rv_plic` with this PR. What do you think?
> 
> yeah that sounds good. Are you guys using our `topgen.py` directly? Because try as we did..i think `topgen` is actually going to end up being very specific to each design. And there are also some issues with `topgen` today.

Our project has a separated `topgen.py` and we build the design out of Open Titan tree. However, we still want to keep our `topgen.py` fork as close to the OT `topgen.py` as possible so we don't need to resolve too many merge conflicts every time we rebase.

**Comment by tjaychen** [comment]
ah sounds good.
if you guys want to discuss / make suggestions on topgen, i think we'd be
happy to discuss also.
there are some issues with topgen especially on inter-module
dependencies that we have not REALLY worked out yet.

**Comment by hcindyl** [comment]
Glad to collaborate. My previous PRs were proposed to support the out-of-tree build. I may have 1-2 more PR(s) after this to support the out-of-tree top level build. If there is any inter-dependency that you would like to have extra set of eyes to resolve, feel free to let us know.

**Comment by tjaychen** [comment]
#2083 actually describes a lot of the issues with topgen.
i also left some examples and long comments in the code itself to hopefully clarify [some of this](https://github.com/lowRISC/opentitan/blob/master/util/topgen.py#L1051)

I think addressing this long term in a way that it could be applied to opentitan and any derivative project in a scaleable way is what we want to achieve.

**Comment by hcindyl** [comment]
Try to fix  license checker again with https://github.com/lowRISC/opentitan/pull/11927

**Comment by rswarbrick** [comment]
Sorry for the slow reply. Yes, passing the name rather than a suffix sounds like a good idea to me.

File: util/ipgen/renderer.py [hunk 1]
```diff
@@ -123,7 +123,10 @@ def _tplfunc_instance_vlnv(self, template_vlnv_str: str) -> str:
         # For example, a core name `rv_plic_component` will result in an
         # instance name 'my_instance_component' (instead of
         # 'my_instance_rv_plic_component').
-        if template_core_name.startswith(self.ip_template.name):
+        if "module_instance_name" in self.ip_config.param_values:
+            idx = len(self.ip_config.param_values["module_instance_name"])
+            template_core_name = template_core_name[idx:]
```
Comment (rswarbrick) [inline util/ipgen/renderer.py:128]: I think this probably also needs an assertion to check that `template_core_name` starts with `self.ip_config.param_values["module_instance_name"]`?

File: util/topgen.py [hunk 2]
```diff
@@ -67,7 +67,10 @@ def ipgen_render(template_name: str, topname: str, params: Dict,
 
     Aborts the program execution in case of an error.
     """
-    instance_name = f"top_{topname}_{template_name}"
+    module_name = template_name
+    if "module_instance_name" in params:
+        module_name = params["module_instance_name"]
```
Comment (rswarbrick) [inline util/topgen.py:72]: Simpler: `module_name = params.get('module_instance_name', template_name)`

**Review by rswarbrick — COMMENTED** [review]
This looks like a good idea to me!

File: util/topgen.py (hunk 2 above)
Comment (hcindyl) [reply util/topgen.py:72]: Thanks for the suggestion. Done.

File: util/ipgen/renderer.py (hunk 1 above)
Comment (hcindyl) [reply util/ipgen/renderer.py:128]: It should be the case from the template, but adding a check here is definitely useful.

**Comment by tjaychen** [comment]
this looks good to me overall. 
@hcindyl would you mind updating the PR / commit description a bit? since it seems a bit out of date now (no longer suffix and actually module name)

**Review by tjaychen — APPROVED** [review]

**Comment by hcindyl** [comment]
> 

It was updated in the commit message. I will update the PR title. :)

**Comment by tjaychen** [comment]
> > 
> 
> It was updated in the commit message. I will update the PR title. :)

thank you! i'll merge when CI passes.

**Pull request merged**